core/domain/html_validation_service.py

```diff
@@ -957,24 +957,39 @@ def get_filename_with_dimensions(exploration_fs, old_filename):
     Args:
         exploration_fs: GcsFileSystem. File system of the Exploration entity.
         old_filename: str. Name of the file whose dimensions need to be
-            calculated.
+            calculated. The filename is not expected to contain the pattern
+            _height_<number>_width_<number>.
 
     Returns:
         str. The new filename of the image file.
+
+    Raises:
+        Exception: The old_filename provided contains the pattern
+            _height_<number>_width_<number>.
     """
     subdirectory = 'image'
     filepath = '%s/%s' % (subdirectory, old_filename)
     try:
-        dot_index = old_filename.rfind('.')
-        filename_regex = r'%s/%s_height_[0-9]+_width_[0-9]+\.\w+' % (
-            subdirectory, old_filename[:dot_index])
-        # Check if the data store contains the updated filename.
-        # If it does, return that filename.
-        for filename in exploration_fs.listdir(
-                subdirectory.encode(encoding='utf-8')):
-            if re.match(filename_regex, filename):
-                (_, new_filename) = filename.split('/')
-                return new_filename
+        # If the old_filename does not contain the pattern
+        # _height_<number>_width_<number>, it is assumed that the
+        # old_filename was not updated. In this case, the following code
+        # iterates over the image subdirectory in the datastore and
+        # looks for a filename with the pattern
+        # <old_filename>_height_<number>_width_<number>.<extension>.
+        # If such a filename is found, it is returned.
+        if re.match(r'^((?!_height_\d+_width_\d+).)*$', old_filename):
```

Comment:
Actually I just realized this is structured wrongly.

At the outset of this function, outside the try/except, do the check to ensure the filename is what you expect it to be. If it isn't, then raise an exception early.

Then do the main logic, since from then on you can assume that the filename is valid.

Note this as a general principle, too -- in general, check for small errors and exit early so that you don't need unnecessary nested if/else conditions.